# Spotify Songs Dataset Analysis

**Author:** Ghaniyul Amri Caulava  
**Module:** Hands On Activity  
**Objective:** Analyze Spotify songs dataset using Pandas and NumPy.

In [1]:
import pandas as pd
import numpy as np

## Dataset Loading

In [2]:
df = pd.read_csv("spotify_songs.csv")
df.head()

,track_id,track_name,track_artist,track_popularity,track_album_id,track_album_name,track_album_release_date,playlist_name,playlist_id,playlist_genre,...,key,loudness,mode,speechiness,acousticness,instrumentalness,liveness,valence,tempo,duration_ms
0,6f807x0ima9a1j3VPbc7VN,I Don't Care (with Justin Bieber) - Loud Luxur...,Ed Sheeran,66,2oCs0DGTsRO98Gh5ZSl2Cx,I Don't Care (with Justin Bieber) [Loud Luxury...,2019-06-14,Pop Remix,37i9dQZF1DXcZDD7cfEKhW,pop,...,6,-2.634,1,0.0583,0.1020,0.000000,0.0653,0.518,122.036,194754
1,0r7CVbZTWZgbTCYdfa2P31,Memories - Dillon Francis Remix,Maroon 5,67,63rPSO264uRjW1X5E6cWv6,Memories (Dillon Francis Remix),2019-12-13,Pop Remix,37i9dQZF1DXcZDD7cfEKhW,pop,...,11,-4.969,1,0.0373,0.0724,0.004210,0.3570,0.693,99.972,162600
2,1z1Hg7Vb0AhHDiEmnDE79l,All the Time - Don Diablo Remix,Zara Larsson,70,1HoSmj2eLcsrR0vE9gThr4,All the Time (Don Diablo Remix),2019-07-05,Pop Remix,37i9dQZF1DXcZDD7cfEKhW,pop,...,1,-3.432,0,0.0742,0.0794,0.000023,0.1100,0.613,124.008,176616
3,75FpbthrwQmzHlBJLuGdC7,Call You Mine - Keanu Silva Remix,The Chainsmokers,60,1nqYsOef1yKKuGOVchbsk6,Call You Mine - The Remixes,2019-07-19,Pop Remix,37i9dQZF1DXcZDD7cfEKhW,pop,...,7,-3.778,1,0.1020,0.0287,0.000009,0.2040,0.277,121.956,169093
4,1e8PAfcKUYoKkxPhrHqw4x,Someone You Loved - Future Humans Remix,Lewis Capaldi,69,7m7vv9wlQ4i0LFuJiE2zsQ,Someone You Loved (Future Humans Remix),2019-03-05,Pop Remix,37i9dQZF1DXcZDD7cfEKhW,pop,...,1,-4.672,1,0.0359,0.0803,0.000000,0.0833,0.725,123.976,189052


In [3]:
df.shape

(32833, 23)

## Dataset Shape
The dataset contains 32,833 rows and 23 columns, indicating a relatively large dataset suitable for exploratory analysis.

In [4]:
pd.DataFrame(df.dtypes, columns=["Data Type"])

,Data Type
track_id,str
track_name,str
track_artist,str
track_popularity,int64
track_album_id,str
track_album_name,str
track_album_release_date,str
playlist_name,str
playlist_id,str
playlist_genre,str


## Missing Value Analysis

In [5]:
df.isnull().sum()

track_id                    0
track_name                  5
track_artist                5
track_popularity            0
track_album_id              0
track_album_name            5
track_album_release_date    0
playlist_name               0
playlist_id                 0
playlist_genre              0
playlist_subgenre           0
danceability                0
energy                      0
key                         0
loudness                    0
mode                        0
speechiness                 0
acousticness                0
instrumentalness            0
liveness                    0
valence                     0
tempo                       0
duration_ms                 0
dtype: int64

In [6]:
missing_percentage = df.isnull().mean() * 100
missing_percentage

track_id                    0.000000
track_name                  0.015229
track_artist                0.015229
track_popularity            0.000000
track_album_id              0.000000
track_album_name            0.015229
track_album_release_date    0.000000
playlist_name               0.000000
playlist_id                 0.000000
playlist_genre              0.000000
playlist_subgenre           0.000000
danceability                0.000000
energy                      0.000000
key                         0.000000
loudness                    0.000000
mode                        0.000000
speechiness                 0.000000
acousticness                0.000000
instrumentalness            0.000000
liveness                    0.000000
valence                     0.000000
tempo                       0.000000
duration_ms                 0.000000
dtype: float64

In [7]:
missing_percentage[missing_percentage > 20]

Series([], dtype: float64)

## Duplicate Handling

In [8]:
df.duplicated().sum()

np.int64(0)

In [9]:
df.duplicated(subset=["track_id"]).sum()

np.int64(4477)

In [10]:
df = df.drop_duplicates(subset=["track_id"])

In [11]:
df.duplicated(subset=["track_id"]).sum()

np.int64(0)

## Statistical Summary

In [12]:
numeric_cols = df.select_dtypes(include=np.number).columns

numeric_cols

Index(['track_popularity', 'danceability', 'energy', 'key', 'loudness', 'mode',
       'speechiness', 'acousticness', 'instrumentalness', 'liveness',
       'valence', 'tempo', 'duration_ms'],
      dtype='str')

In [13]:
summary = pd.DataFrame()

for col in numeric_cols:

    mean = np.mean(df[col])

    median = np.median(df[col])

    std = np.std(df[col])

    q1 = np.percentile(df[col], 25)

    q3 = np.percentile(df[col], 75)

    iqr = q3 - q1

    summary.loc[col, "Mean"] = mean
    summary.loc[col, "Median"] = median
    summary.loc[col, "Std"] = std
    ssummary.loc[col, "IQR"] = np.percentile(df[col], 75) - np.percentile(df[col], 25)

summary

NameError: name 'ssummary' is not defined

## Statistical Summary Insight
The statistical summary shows that song characteristics vary substantially across the dataset. Features such as tempo and loudness have relatively large dispersion, indicating broad variation in musical production styles.

In [ ]:
genre_stats = df.groupby("playlist_genre")[
    ["track_popularity", "danceability", "energy", "valence"]
].agg(["mean", "std", "median"])

genre_stats

track_popularity                   danceability            \
                           mean        std median         mean       std   
playlist_genre                                                             
edm                   30.678286  20.346961   33.0     0.657639  0.123722   
latin                 41.439691  23.394529   45.0     0.711012  0.117222   
pop                   45.905300  24.616386   50.0     0.637698  0.128997   
r&b                   35.929396  23.662984   38.0     0.667475  0.137610   
rap                   41.822811  22.765566   46.0     0.715991  0.136217   
rock                  39.694309  24.229616   44.0     0.518519  0.140418   

                         energy                    valence                   
               median      mean       std median      mean       std median  
playlist_genre                                                               
edm             0.660  0.809604  0.136550  0.838  0.397491  0.228631  0.365  
latin           0.727  0.710468  0.156036  0.733  0.607390  0.225631  0.632  
pop             0.650  0.701031  0.172949  0.727  0.502176  0.221924  0.499  
r&b             0.687  0.588932  0.181949  0.594  0.537936  0.225879  0.548  
rap             0.734  0.649832  0.172107  0.666  0.505182  0.225269  0.517  
rock            0.522  0.733067  0.197484  0.779  0.532560  0.230204  0.526

## Genre Distribution Analysis
Genre-based aggregation was performed to examine how popularity and audio characteristics differ across music genres.

In [ ]:
variance = df.groupby("playlist_genre")["track_popularity"].var()

variance

playlist_genre
edm      413.998817
latin    547.303966
pop      605.966474
r&b      559.936831
rap      518.271006
rock     587.074282
Name: track_popularity, dtype: float64

In [ ]:
variance.idxmax()

'pop'

In [ ]:
variance.max()

np.float64(605.9664735602968)

## Artist Analysis

In [ ]:
top10_artists = df["track_artist"].value_counts().head(10)

top10_artists

track_artist
Queen                        130
Martin Garrix                 87
Don Omar                      84
David Guetta                  81
Dimitri Vegas & Like Mike     68
Hardwell                      68
Drake                         68
The Chainsmokers              66
Logic                         65
Guns N' Roses                 63
Name: count, dtype: int64

In [ ]:
artist_popularity = df.groupby("track_artist")["track_popularity"].mean()

artist_popularity.loc[top10_artists.index]

track_artist
Queen                        42.400000
Martin Garrix                41.402299
Don Omar                     39.059524
David Guetta                 49.370370
Dimitri Vegas & Like Mike    36.220588
Hardwell                     36.323529
Drake                        41.808824
The Chainsmokers             49.227273
Logic                        41.907692
Guns N' Roses                27.206349
Name: track_popularity, dtype: float64

In [ ]:
artist_popularity.loc[top10_artists.index].idxmax()

'David Guetta'

In [ ]:
comparison = pd.DataFrame({
    "Number_of_Tracks": top10_artists,
    "Average_Popularity": artist_popularity.loc[top10_artists.index]
})

comparison

,Number_of_Tracks,Average_Popularity
track_artist,,
Queen,130,42.400000
Martin Garrix,87,41.402299
Don Omar,84,39.059524
David Guetta,81,49.370370
Dimitri Vegas & Like Mike,68,36.220588
Hardwell,68,36.323529
Drake,68,41.808824
The Chainsmokers,66,49.227273
Logic,65,41.907692


In [ ]:
comparison = pd.DataFrame({
    "Number_of_Tracks": top10_artists,
    "Average_Popularity": artist_popularity.loc[top10_artists.index]
})

comparison

,Number_of_Tracks,Average_Popularity
track_artist,,
Queen,130,42.400000
Martin Garrix,87,41.402299
Don Omar,84,39.059524
David Guetta,81,49.370370
Dimitri Vegas & Like Mike,68,36.220588
Hardwell,68,36.323529
Drake,68,41.808824
The Chainsmokers,66,49.227273
Logic,65,41.907692


## Track Filtering Analysis

After applying multiple filtering criteria, 579 tracks satisfied all conditions related to popularity, danceability, energy, and duration.

The pop genre appears most frequently among these filtered tracks, suggesting that pop music tends to balance commercial popularity with energetic and dance-friendly audio characteristics.

This may explain why pop music often performs strongly on mainstream streaming platforms.

In [ ]:
filtered_tracks = df[
    (df["track_popularity"] > 70) &
    (df["danceability"] > 0.7) &
    (df["energy"] > 0.6) &
    (df["duration_ms"] < 240000)
]

In [ ]:
len(filtered_tracks)

579

In [ ]:
filtered_tracks["playlist_genre"].value_counts()

playlist_genre
pop      193
latin    173
rap      141
r&b       42
edm       19
rock      11
Name: count, dtype: int64

In [ ]:
filtered_tracks["playlist_genre"].value_counts().idxmax()

'pop'

## NumPy Normalization

The nine audio feature columns were extracted as a NumPy matrix with shape (28356, 9).

Min-max normalization was successfully applied using vectorized NumPy operations, transforming every feature into a range between 0 and 1.

This normalization ensures that all audio features contribute equally to subsequent mathematical analysis.

In [ ]:
features = [
    "danceability",
    "energy",
    "loudness",
    "speechiness",
    "acousticness",
    "instrumentalness",
    "liveness",
    "valence",
    "tempo"
]


In [ ]:
X = df[features].to_numpy()

X

array([[ 7.48000e-01,  9.16000e-01, -2.63400e+00, ...,  6.53000e-02,
         5.18000e-01,  1.22036e+02],
       [ 7.26000e-01,  8.15000e-01, -4.96900e+00, ...,  3.57000e-01,
         6.93000e-01,  9.99720e+01],
       [ 6.75000e-01,  9.31000e-01, -3.43200e+00, ...,  1.10000e-01,
         6.13000e-01,  1.24008e+02],
       ...,
       [ 5.29000e-01,  8.21000e-01, -4.89900e+00, ...,  1.50000e-01,
         4.36000e-01,  1.27989e+02],
       [ 6.26000e-01,  8.88000e-01, -3.36100e+00, ...,  3.43000e-01,
         3.08000e-01,  1.28008e+02],
       [ 6.03000e-01,  8.84000e-01, -4.57100e+00, ...,  7.42000e-01,
         8.94000e-02,  1.27984e+02]], shape=(28356, 9))

In [ ]:
X.shape

(28356, 9)

In [ ]:
X_scaled = (X - X.min(axis=0)) / (X.max(axis=0) - X.min(axis=0))

In [ ]:
X_scaled.shape

(28356, 9)

In [ ]:
X_scaled.min(axis=0)

array([0., 0., 0., 0., 0., 0., 0., 0., 0.])

In [ ]:
X_scaled.max(axis=0)

array([1., 1., 1., 1., 1., 1., 1., 1., 1.])

## Correlation Matrix Analysis

The strongest positive correlation was found between **energy** and **loudness** with a correlation value of approximately **0.68**.

This suggests that songs with higher energy levels tend to be produced with higher loudness levels. Musically, energetic songs such as dance or electronic tracks are commonly mastered louder to create a stronger impact on listeners.

The strongest negative correlation was found between **energy** and **acousticness** with a correlation value of approximately **-0.55**.

This indicates that songs with high energy are generally less acoustic in nature. In musical terms, highly energetic tracks often rely more on electronic production elements, while acoustic songs tend to be softer and less intense.

In [ ]:
corr_matrix = np.corrcoef(X_scaled, rowvar=False)

corr_matrix

array([[ 1.        , -0.08143602,  0.01529721,  0.18345308, -0.02887824,
        -0.00226669, -0.12700236,  0.33372863, -0.18457517],
       [-0.08143602,  1.        ,  0.68213774, -0.0290084 , -0.54588619,
         0.02381792,  0.16370922,  0.14971046,  0.15154451],
       [ 0.01529721,  0.68213774,  1.        ,  0.01298101, -0.37160236,
        -0.15430904,  0.08192704,  0.04951016,  0.09670611],
       [ 0.18345308, -0.0290084 ,  0.01298101,  1.        ,  0.02494542,
        -0.10795933,  0.05934289,  0.06470092,  0.0327286 ],
       [-0.02887824, -0.54588619, -0.37160236,  0.02494542,  1.        ,
        -0.0031005 , -0.07453962, -0.01899856, -0.11433521],
       [-0.00226669,  0.02381792, -0.15430904, -0.10795933, -0.0031005 ,
         1.        , -0.00850485, -0.17416318,  0.02148405],
       [-0.12700236,  0.16370922,  0.08192704,  0.05934289, -0.07453962,
        -0.00850485,  1.        , -0.01992507,  0.02202664],
       [ 0.33372863,  0.14971046,  0.04951016,  0.06470092, -0

In [ ]:
corr_df = pd.DataFrame(
    corr_matrix,
    index=features,
    columns=features
)

corr_df

,danceability,energy,loudness,speechiness,acousticness,instrumentalness,liveness,valence,tempo
danceability,1.000000,-0.081436,0.015297,0.183453,-0.028878,-0.002267,-0.127002,0.333729,-0.184575
energy,-0.081436,1.000000,0.682138,-0.029008,-0.545886,0.023818,0.163709,0.149710,0.151545
loudness,0.015297,0.682138,1.000000,0.012981,-0.371602,-0.154309,0.081927,0.049510,0.096706
speechiness,0.183453,-0.029008,0.012981,1.000000,0.024945,-0.107959,0.059343,0.064701,0.032729
acousticness,-0.028878,-0.545886,-0.371602,0.024945,1.000000,-0.003101,-0.074540,-0.018999,-0.114335
instrumentalness,-0.002267,0.023818,-0.154309,-0.107959,-0.003101,1.000000,-0.008505,-0.174163,0.021484
liveness,-0.127002,0.163709,0.081927,0.059343,-0.074540,-0.008505,1.000000,-0.019925,0.022027
valence,0.333729,0.149710,0.049510,0.064701,-0.018999,-0.174163,-0.019925,1.000000,-0.025135
tempo,-0.184575,0.151545,0.096706,0.032729,-0.114335,0.021484,0.022027,-0.025135,1.000000


In [ ]:
corr_no_diag = corr_matrix.copy()

np.fill_diagonal(corr_no_diag, 0)

In [ ]:
max_corr_index = np.unravel_index(
    np.argmax(corr_no_diag),
    corr_no_diag.shape
)

max_corr_index

(np.int64(2), np.int64(1))

In [ ]:
features[max_corr_index[0]], features[max_corr_index[1]]

('loudness', 'energy')

In [ ]:
corr_no_diag[max_corr_index]

np.float64(0.6821377424910231)

In [ ]:
min_corr_index = np.unravel_index(
    np.argmin(corr_no_diag),
    corr_no_diag.shape
)

min_corr_index

(np.int64(1), np.int64(4))

In [ ]:
features[min_corr_index[0]], features[min_corr_index[1]]

('energy', 'acousticness')

In [ ]:
corr_no_diag[min_corr_index]

np.float64(-0.5458861907892345)

## Boolean Masking Analysis

Using NumPy boolean masking, 5,625 songs were identified as having energy values above one standard deviation from the mean.

The average popularity of this subset was **36.12**, while the average popularity of the full dataset was **42.48**.

This indicates that songs with extremely high energy levels do not necessarily achieve higher popularity.

One possible explanation is that highly energetic songs appeal strongly to specific listener groups, but mainstream listeners may prefer a wider variety of musical styles.

In [ ]:
energy = X[:, 1]

In [ ]:
mean_energy = np.mean(energy)

std_energy = np.std(energy)

mean_energy, std_energy

(np.float64(0.6983875199252364), np.float64(0.18349980900492102))

In [ ]:
mask = energy > (mean_energy + std_energy)

In [ ]:
mask.sum()

np.int64(4853)

In [ ]:
high_energy_tracks = df[mask]

In [ ]:
subset_popularity = high_energy_tracks["track_popularity"].mean()

subset_popularity

np.float64(34.02884813517412)

In [ ]:
overall_popularity = df["track_popularity"].mean()

overall_popularity

np.float64(39.329771476936095)

In [ ]:
subset_popularity, overall_popularity

(np.float64(34.02884813517412), np.float64(39.329771476936095))

# Bonus 1 – Artist Audio Fingerprint

A fingerprint vector was created for each artist by computing the mean value of nine normalized audio features.

Cosine similarity was implemented manually using NumPy based on the mathematical definition without external machine learning libraries.

Three artist pairs were tested:

- Queen vs Martin Garrix = 0.9996  
- Don Omar vs Guns N' Roses = 1.0000  
- Drake vs Hardwell = 0.9996  

Surprisingly, all similarity values were extremely high, even for artists from very different genres.

This suggests that averaging audio features at the artist level reduces stylistic variation, causing artist vectors to become very similar in direction.

As a result, cosine similarity in this form does not strongly separate artists by genre.

This experiment demonstrates an important data science lesson: a mathematically correct model does not always produce highly discriminative results, and feature engineering choices strongly influence analytical outcomes.

In [ ]:
scaled_df = df.copy()

scaled_df[features] = X_scaled

In [ ]:
scaled_df.shape

(28356, 23)

In [ ]:
def artist_fingerprint(df, artist_name):
    
    artist_data = df[df["track_artist"] == artist_name]
    
    mean_vector = artist_data[features].mean().to_numpy()
    
    return mean_vector

In [ ]:
def cosine_similarity(A, B):
    
    dot_product = np.dot(A, B)
    
    norm_A = np.linalg.norm(A)
    
    norm_B = np.linalg.norm(B)
    
    return dot_product / (norm_A * norm_B)

In [ ]:
pairs = [
    ("Queen", "Martin Garrix"),
    ("Don Omar", "Guns N' Roses"),
    ("Drake", "Hardwell")
]

for artist1, artist2 in pairs:
    
    vec1 = artist_fingerprint(scaled_df, artist1)
    
    vec2 = artist_fingerprint(scaled_df, artist2)
    
    similarity = cosine_similarity(vec1, vec2)
    
    print(f"{artist1} vs {artist2}: {similarity:.4f}")

Queen vs Martin Garrix: 0.9527
Don Omar vs Guns N' Roses: 0.9553
Drake vs Hardwell: 0.9579


# Final Conclusion

This analysis demonstrates that Spotify song characteristics can be effectively explored using Pandas and NumPy.

Several key findings were identified:

- Pop music shows the highest popularity variance.
- High-energy songs tend to be louder and less acoustic.
- A larger number of songs does not necessarily guarantee higher popularity.
- Extremely energetic songs do not always achieve higher audience reception.
- Artist fingerprint similarity suggests that feature averaging can reduce stylistic distinction.

Overall, this project highlights how mathematical analysis can uncover meaningful patterns in music data and support data-driven decision making in recommendation systems.

# AI Usage Disclosure

Artificial Intelligence tools were used during this assignment as a learning assistant to:

- understand NumPy mathematical implementations
- debug Python and Jupyter Notebook execution errors
- improve code structure and notebook organization
- assist in interpreting analytical results

All code execution, testing, debugging, and final validation were performed manually by the author.

AI was used as a supporting learning tool, not as a replacement for understanding the concepts.